packages/langium/src/utils/cst-util.ts

```diff
@@ -155,30 +155,53 @@ export function isCommentNode(cstNode: CstNode, commentNames: string[]): boolean
     return isLeafCstNode(cstNode) && commentNames.includes(cstNode.tokenType.name);
 }
 
+/**
+ * Finds the leaf CST node at the specified 0-based string offset.
+ * If no CST node exists at the specified position, it will return the leaf node before it.
+ *
+ * @param node The CST node to search through
+ * @param offset The specified offset
+ * @returns The CST node closest to the specified output.
+ */
 export function findLeafNodeAtOffset(node: CstNode, offset: number): LeafCstNode | undefined {
     if (isLeafCstNode(node)) {
         return node;
     } else if (isCompositeCstNode(node)) {
-        let firstChild = 0;
-        let lastChild = node.content.length - 1;
-        while (firstChild < lastChild) {
-            const middleChild = Math.floor((firstChild + lastChild) / 2);
-            const n = node.content[middleChild];
-            if (n.offset > offset) {
-                lastChild = middleChild - 1;
-            } else if (n.end <= offset) {
-                firstChild = middleChild + 1;
-            } else {
-                return findLeafNodeAtOffset(n, offset);
-            }
-        }
-        if (firstChild === lastChild) {
-            return findLeafNodeAtOffset(node.content[firstChild], offset);
+        const searchResult = binarySearch(node, offset);
+        if (searchResult) {
+            return findLeafNodeAtOffset(searchResult, offset);
```

Comment:
Code looks very good to me, I'm just surprised for the new test in [cst-utils.test.ts](https://github.com/eclipse-langium/langium/pull/1267/files#diff-5d9ca30c93ce56e1c0049557d3bf3df7e9f66ee91a290239f3baaf036efc325f) it runs into a recursion of depth 12 😳 
Would it make sense to add a special treatment for case of `node.content.length === 1`? 

<img width="422" alt="image" src="https://github.com/eclipse-langium/langium/assets/8513883/d235c605-fd7b-4a20-b17c-1e5e182f819f">

Reply:
The recursion depth is exactly the same depth as the CST depth at the specific position, since it's a tree based recursion algorithm. We could probably improve that by making the function work iteratively instead of recursively. Though that might be something for a separate PR.

> Would it make sense to add a special treatment for case of node.content.length === 1?

I've tried that just now, but the amount of comparisons we need is actually larger than without it as we have a lot of cases where `node.content.length > 1`. It only makes the method more verbose in my opinion.

Reply:
>  It only makes the method more verbose in my opinion.
That's probably true.

I just though that this method is a prominent one and performance could be a relevant topic in case of giant input files. The recursion depth is maybe not reducible, but calling `binarySearch` and doing one or two iterations in its loop makes little sense in case there's just a single child.

Reply:
Performance is relevant, but the amount of comparisons we're doing for n = 1 is exactly the same if we put a `if (n = 1)` in the front of the method, or whether we just let the method run into the loop. In fact, it leads to more comparison due to the new `if (n = 1)` at the start of the method. We still need to perform the "in bounds" checking of the child node even if its the only child.

Reply:
Makes sense.
> We still need to perform the "in bounds" checking of the child node even if its the only child.

I wasn't sure on which assumptions we can rely at this point.